File: app/src/main/java/org/neidhardt/dynamicsoundboard/navigationdrawer/soundsheets/views/SoundSheetsPresenter.java
package org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.views;

import android.view.View;
import org.neidhardt.dynamicsoundboard.dao.SoundSheet;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.navigationdrawer.views.NavigationDrawerListPresenter;
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataAccess;
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataStorage;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.events.OpenSoundSheetEvent;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.model.SoundSheetsDataAccess;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.model.SoundSheetsDataStorage;

import java.util.ArrayList;
import java.util.List;

/**
 * File created by eric.neidhardt on 26.05.2015.
 */
public class SoundSheetsPresenter
		extends
			NavigationDrawerListPresenter<SoundSheets>
		implements
			SoundSheetsAdapter.OnItemClickListener
{
	private SoundSheetsDataAccess soundSheetsDataAccess;
	private SoundSheetsDataStorage soundSheetsDataStorage;
	private SoundsDataAccess soundsDataAccess;
	private SoundsDataStorage soundsDataStorage;

	private SoundSheetsAdapter adapter;

	public SoundSheetsPresenter(SoundSheetsDataAccess soundSheetsDataAccess,
								SoundSheetsDataStorage soundSheetsDataStorage,
								SoundsDataAccess soundsDataAccess,
								SoundsDataStorage soundsDataStorage)
	{
		this.soundSheetsDataAccess = soundSheetsDataAccess;
		this.soundSheetsDataStorage = soundSheetsDataStorage;
		this.soundsDataAccess = soundsDataAccess;
		this.soundsDataStorage = soundsDataStorage;
	}

	@Override
	protected boolean isEventBusSubscriber()
	{
		return false;
	}

	@Override
	public void deleteSelectedItems()
	{
		List<SoundSheet> soundSheetsToRemove = this.getSoundSheetsSelectedForDeletion();
		for (SoundSheet soundSheet: soundSheetsToRemove)
		{
			List<EnhancedMediaPlayer> soundsInFragment = this.soundsDataAccess.getSoundsInFragment(soundSheet.getFragmentTag());
			this.soundsDataStorage.removeSounds(soundsInFragment);
		}
		this.soundSheetsDataStorage.removeSoundSheets(soundSheetsToRemove);

		super.onSelectedItemsDeleted();
	}

	@Override
	public void onItemClick(View view, SoundSheet data, int position)
	{
		if (this.isInSelectionMode())
		{
			data.setIsSelectedForDeletion(!data.getIsSelectedForDeletion());
			super.onItemSelectedForDeletion();
		}
		else
		{
			this.soundSheetsDataAccess.setSelectedItem(position);
			this.getEventBus().post(new OpenSoundSheetEvent(data));
		}
		this.adapter.notifyItemChanged(position);
	}

	@Override
	protected int getNumberOfItemsSelectedForDeletion()
	{
		return this.getSoundSheetsSelectedForDeletion().size();
	}

	@Override
	protected void deselectAllItemsSelectedForDeletion()
	{
		List<SoundSheet> selectedSoundSheets = this.getSoundSheetsSelectedForDeletion();
		for (SoundSheet soundSheet : selectedSoundSheets)
		{
			soundSheet.setIsSelectedForDeletion(false);
			this.adapter.notifyItemChanged(soundSheet);
		}
	}

	private List<SoundSheet> getSoundSheetsSelectedForDeletion()
	{
		List<SoundSheet> selectedSoundSheets = new ArrayList<>();
		List<SoundSheet> existingSoundSheets = this.adapter.getValues();
		for(SoundSheet soundSheet : existingSoundSheets)
		{
			if (soundSheet.getIsSelectedForDeletion())
				selectedSoundSheets.add(soundSheet);
		}
		return selectedSoundSheets;
	}

	public void setAdapter(SoundSheetsAdapter adapter)
	{
		this.adapter = adapter;
	}

	public List<SoundSheet> getSoundSheets()
	{
		return this.soundSheetsDataAccess.getSoundSheets();
	}

	public List<EnhancedMediaPlayer> getSoundsInFragment(String fragmentTag)
	{
		return this.soundsDataAccess.getSoundsInFragment(fragmentTag);
	}

	public SoundSheetsDataAccess getSoundSheetsDataAccess()
	{
		return soundSheetsDataAccess;
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/navigationdrawer/soundsheets/SoundSheetsPresenter.kt
package org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets

import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.navigationdrawer.NavigationDrawerListBasePresenter
import org.neidhardt.dynamicsoundboard.model.SoundSheet
import kotlin.properties.Delegates

/**
 * File created by eric.neidhardt on 26.05.2015.
 */


open class SoundSheetsPresenter() : NavigationDrawerListBasePresenter() {

	companion object {
		fun createSoundSheetPresenter(adapter: SoundSheetsAdapter): SoundSheetsPresenter {
			return SoundSheetsPresenter().apply {
				this.adapter = adapter
			}
		}
	}

	private val soundSheetManager = SoundboardApplication.soundSheetManager
	private val soundManager = SoundboardApplication.soundManager

	var adapter: SoundSheetsAdapter by Delegates.notNull<SoundSheetsAdapter>()
	protected val values: List<SoundSheet> get() = this.soundSheetManager.soundSheets

	override fun onAttachedToWindow() {
		this.adapter.notifyDataSetChanged()
	}

	override fun onDetachedFromWindow() { } // nothing to do

	override fun deleteSelectedItems() {
		val soundSheetsToRemove = this.getSoundSheetsSelectedForDeletion()
		for (soundSheet in soundSheetsToRemove) {
			// remove all sounds of this soundSheet to free resources
			this.soundManager.sounds[soundSheet]?.let {
				this.soundManager.remove(soundSheet, it)
			}
		}
		this.soundSheetManager.remove(soundSheetsToRemove)
		this.stopDeletionMode()
	}

	fun onItemClick(data: SoundSheet) {
		if (this.isInSelectionMode) {
			data.isSelectedForDeletion = !data.isSelectedForDeletion
			this.adapter.notifyItemChanged(data)
		}
		else {
			this.soundSheetManager.setSelected(data)
		}
	}

	public override val numberOfItemsSelectedForDeletion: Int
		get() = this.getSoundSheetsSelectedForDeletion().size

	override val itemCount: Int
		get() = this.values.size

	override fun deselectAllItemsSelectedForDeletion() {
		val selectedSoundSheets = this.getSoundSheetsSelectedForDeletion()
		for (soundSheet in selectedSoundSheets) {
			soundSheet.isSelectedForDeletion = false
		}
		this.adapter.notifyDataSetChanged()
	}

	override fun selectAllItems() {
		val selectedSoundSheets = this.values
		for (soundSheet in selectedSoundSheets) {
			soundSheet.isSelectedForDeletion = true
		}
		this.adapter.notifyDataSetChanged()
	}

	private fun getSoundSheetsSelectedForDeletion(): List<SoundSheet> {
		val existingSoundSheets = this.values
		val selectedSoundSheets = existingSoundSheets.filter { it.isSelectedForDeletion }
		return selectedSoundSheets
	}

}
